nucypher-core/src/dkg.rs

```diff
@@ -21,6 +109,250 @@ pub enum FerveoVariant {
+        #[inline]
+        fn as_ref(&self) -> &[u8] {
+            self.0.as_bytes()
+        }
+    }
+
+    impl fmt::Display for RequestSharedSecret {
+        /// Format shared secret information.
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            write!(f, "RequestSharedSecret...")
+        }
+    }
+
+    /// A request public key.
+    #[derive(PartialEq, Eq, Hash, Copy, Clone, Debug, Serialize, Deserialize)]
+    pub struct RequestPublicKey(pub(crate) PublicKey);
+
+    /// Implementation of request public key
+    impl RequestPublicKey {
+        /// Convert this public key to a byte array.
+        #[inline]
+        pub fn to_bytes(&self) -> [u8; 32] {
+            self.0.to_bytes()
+        }
+
+        /// View this public key as a byte array.
+        #[inline]
+        pub fn as_bytes(&self) -> &[u8; 32] {
+            self.0.as_bytes()
+        }
+    }
+
+    impl AsRef<[u8]> for RequestPublicKey {
+        /// View this public key as a byte array.
+        #[inline]
+        fn as_ref(&self) -> &[u8] {
+            self.0.as_bytes()
+        }
+    }
+
+    impl fmt::Display for RequestPublicKey {
+        /// Format public key information.
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            write!(f, "RequestPublicKey: {}", hex::encode(&self.as_ref()[..8]))
+        }
+    }
+
+    impl<'a> From<&'a RequestSecretKey> for RequestPublicKey {
+        /// Compute corresponding [`RequestPublicKey`].
+        fn from(secret: &'a RequestSecretKey) -> RequestPublicKey {
+            let public_key = PublicKey::from(&secret.0);
+            RequestPublicKey(public_key)
+        }
+    }
+
+    impl<'a> ProtocolObjectInner<'a> for RequestPublicKey {
+        fn version() -> (u16, u16) {
+            (1, 0)
+        }
+
+        fn brand() -> [u8; 4] {
+            *b"TRPk"
+        }
+
+        fn unversioned_to_bytes(&self) -> Box<[u8]> {
+            messagepack_serialize(&self)
+        }
+
+        fn unversioned_from_bytes(
+            minor_version: u16,
+            bytes: &[u8],
+        ) -> Option<Result<Self, String>> {
+            if minor_version == 0 {
+                Some(messagepack_deserialize(bytes))
+            } else {
+                None
+            }
+        }
+    }
+
+    impl<'a> ProtocolObject<'a> for RequestPublicKey {}
+
+    /// A request secret key.
+    #[derive(ZeroizeOnDrop)]
+    pub struct RequestSecretKey(pub(crate) StaticSecret);
+
+    impl RequestSecretKey {
+        /// Perform diffie-hellman
+        pub fn derive_shared_secret(
+            &self,
+            their_public_key: &RequestPublicKey,
+        ) -> RequestSharedSecret {
+            let shared_secret = self.0.diffie_hellman(&their_public_key.0);
+            RequestSharedSecret(shared_secret)
+        }
+
+        /// Create secret key from rng.
+        pub fn random_from_rng<T: RngCore + CryptoRng>(csprng: T) -> Self {
```

Comment:
I don't know if you have particularly strong feelings about the name, but Umbral uses `_with_rng()` suffixes for such methods.

Also, you can write the method declaration as `fn random_from_rng(csprng: &(impl RngCore + CryptoRng))` (or even `fn random_from_rng(csprng: &impl CryptoRngCore)` using the `CryptoRngCore` alias introduced in `rand_core=0.6.4`)

Reply:
✔️